api/envoy/config/listener/v3/listener.proto

```diff
@@ -119,6 +119,40 @@ message Listener {
   // :ref:`FAQ entry <faq_how_to_setup_sni>`.
   repeated FilterChain filter_chains = 3;
 
+  // A list of filter chain matching predicate extension type URLs specifying
+  // the order of selection for the most specific filter chain. At each step,
+  // the predicate is evaluated for each filter chain. If the predicate returns
+  // false, the filter chain is eliminated from subsequent consideration. If
+  // the predicate returns true, but there is a more specific match, it is also
+  // eliminated.
+  //
+  // For example, for SNI ``www.example.com`` the filter chain with the
+  // predicate ```*.example.com``` remains and the filter chain with the
+  // predicate ```*.com``` is eliminated, assuming there is no filter chain
+  // with the exact SNI match. In general, the order of specificity is domain
+  // specific as defined by the predicate extension.
+  //
+  // In case, there is more than one filter chain candidates remaining after
+  // the process completes, the first of the filter chains in the order of
```

Comment:
Are there cases where this behavior is desirable? This has caused endless pain for Istio and other Envoy users I have talked to. See https://github.com/envoyproxy/envoy/issues/12572 for a lot of discussion.

Perhaps this is an opportunity to allow opting into a behavior that makes the matching more similar to other matching systems and that is easier to use without having a bunch of duplicated filter chains to appease the matching system?

Reply:
@howardjohn The structured match is important for nested trie-matching algorithm, I think we cannot drop that at this point. The specific issue you had is with the defaulting. Each extension should default to allow-all, not deny-all, like in the above example issue you linked. We are fixing that with enforcement that empty matching predicate matches everything. The problem then becomes that multiple match chains will overlap, so we are also addressing that with the declared order.

Reply:
We will allow wildcards for ports, transport security, and ALPN, so the first issue is solved. We can also consider complement expression (not-expression), if necessary. Basically, there's nothing wrong with trie as long as the complement fits into the domain well. The idea is to extend the match conditions to avoid duplication of the chains, which makes sense.

Second issue is again similar. We'll just assume any value for path or header matches if not specified, unlike the existing proto. Yes, this was a mistake in the original definition, I think.

Is there something beyond defaulting and more expressive match conditions?

Reply:
> We'll just assume any value for path or header matches if not specified, unlike the existing proto.

I think this may be sufficient, will look into it a bit more